### PR TITLE
Implement UI to switch DBs

### LINKED ISSUE #173: Implement frontend for switching between different databases
## Problem

The user might have multiple databases configured in Mathesar. The UI should allow the user to switch between these databases.

## Proposed solution

We need to allow users to switch between different databases. These databases will be configured while setting up Mathesar and cannot be changed from the UI (at least, not yet), so the user just needs to see what's already configured and switch between them.

## Additional context

Blocked by:
- #163 
- #172

### PR DESCRIPTION
**Pull request by pavish**
Fixes #173 

This PR:
- Implements UI to switch DBs
- Refactors Menu component
   - Removes usage of grid and aligns content manually
   - Adds composable MenuItemWrapper and MenuItemContents components
- Fixes several small alignment issues

Note to reviewers:
- For alignment issues and css related changes, the commit messages should provide the details and purpose of the change
- I've removed the obsolete `App.test.ts` integ test, which failed due to a `Function called outside component initialization` error. It was more trouble to fix than to remove. We'll have to write individual integ tests per page

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by seancolsen — APPROVED** [review]
## Behavior

I played with these changes and I think that most of the front end stuff is working great from the user's perspective. Nicely done!

Some small notes:

- I added 8dfc2349cf1211104518afdaeb58591c25445920 to improve the help popover

    | Before | After |
    | -- | -- |
    | image | image |


- I'm not sure what we are expecting from the backend currently, but I wasn't able to actually get two databases running because within the database I created I didn't see a `public` schema and was not able to create any new schemas. I might be doing something wrong on my end though. Not sure. Were you able to get multiple databases working?

    image

## Menu system

I appreciate the direction you're pushing the design of the Menu system. It really _is_ nice getting rid of CSS grid (and thus `display:contents`)! And it will be helpful to use these components more widely. I can see how, with a little more work, this JS-based approach to alignment would leave us with zero sacrifices over my original grid-based approach. (The small sacrifice at the moment is that we would hypothetically need to specify widths in some cases, but that's not a huge deal.) Awesome!

A couple smaller points...

### Reactivity

Aside from my overall satisfaction with your high-level approach, some of the code in `menu/utils.ts` felt strange to me. I poked at it a bit, first just trying to understand it, and eventually making some adjustments in 9c49edce5bceff1ed616a658dc9fcaeddfd5a589 that I'm hoping you might like.

Before my change:

- Inspecting `MenuItemContents.svelte` gave me the impression that the menu would react to changes from the `icon` prop. But on closer inspection in `menu/utils.ts` I noticed that the reactivity was only implemented for the case of setting the store values from `false` to `true` (but not from `true` to `false`). I wouldn't expect that we'd necessarily have the need for reactively removing an icon from a menu item, but given that this is a component-library component, I think it would be nice for it to behave predictably in that manner. Further, the use-case of conditionally rendering menu items _is_ a relatively common need. Imagine that someone has a menu which is initially mounted while containing an item with a control. The menu will begin with a column for the control cell. Good. But then if that menu item is reactively removed from the menu, that leaves the menu with an empty control column. In b93b949ed7c468c265ce7c0d9016222adc5528f3, I adjusted the component story to demo this bug.

    [before.webm](https://user-images.githubusercontent.com/42411/235031007-87a420ae-cbe6-4714-b618-17ea696ca653.webm)

    This sort of bug could be tricky to track down later, so I think it's worth addressing while we're in the midst of this refactor.

- The reactive logic feels very imperative, with some code duplication too.
- I was confused why the second parameter of `registerMenuItem` was a function.
- `MenuAlignmentStores` felt like an awkward mix of properties relevant to the `Menu` component and properties relevant to each menu item.
- None of these problems are super important, but I got a little curious thinking about how to do this more cleanly.

After my change:

- A new lower-level `ReductionStore` class encapsulates the Menu/MenuItem logic more declaratively and generically.
- A `MenuController` is passed down via context and uses one `ReductionStore` per column.
- Each `MenuItemContents` registers its cells as inputs to the `ReductionStore` instances.
- No more imperative logic. And the reactivity bug is fixed:

    [after.webm](https://user-images.githubusercontent.com/42411/235031026-fd282a92-a237-496d-8180-5ef398ae8961.webm)

I think your code is a tiny bit more performant because the boolean computation is short circuited earlier. But I think my code is slightly more stable and maintainable because the reactivity is more robust and declarative. This situation reminds me of the cell-selection conversation we had via chat recently where I was arguing for greater stability at the cost of performance, with my point being that the performance impact was negligible. We might have a similar predicament here, so I'm curious if you'll have concerns with my commit. Perhaps you already thought of my approach but decided against it. I wouldn't imagine we'd ever have an enormous menu, so I'm not worried about my approach performance wise. But if you don't like it, then please feel free to revert it and merge the PR yourself.


### Padding

I noticed some problems with alignment due to the cell padding.

image

It would be nice to fix this, but I don't think it needs to block this PR since we don't have any menus with inconsistent icon/control usage.

**Comment by seancolsen** [comment]
@pavish I'm done reviewing, but I'll leave it to you to decide if you want to merge as-is or push more changes to revert my reactivity change and/or fix the padding alignment issue.

File: mathesar_ui/src/pages/database/DatabaseNavigationList.svelte
Comment (pavish) [reply mathesar_ui/src/pages/database/DatabaseNavigationList.svelte:19]: I assumed it'd be best to use the MenuHeading component since we might add aria attributes to the common component later to improve accessibility.

**Comment by pavish** [comment]
> I'm not sure what we are expecting from the backend currently, but I wasn't able to actually get two databases running because within the database I created I didn't see a public schema and was not able to create any new schemas. I might be doing something wrong on my end though. Not sure. Were you able to get multiple databases working?

Yes, I tested this in the prod setup and the dev setup. You'll have to clear all existing containers and volumes, add the DBs to the env file, and start it fresh to get things working at the moment. For dev setup, you could also follow the instructions [here](https://github.com/centerofci/mathesar/issues/2006#issuecomment-1507191139).

I've created https://github.com/centerofci/mathesar/issues/2852 to document adding/removing DBs in an existing Mathesar instance.

**Pull request merged**
